README.md
Introduction
============

This Bundle provide barcode rendering using our barcode generator base library https://github.com/paterik/BGBarcodeGenerator
supports ImageMagick/GDLib based image rendering, svg and html table drawn output of the following 1D/2D bar codes:

1d barcodes
-----------
* C39 : CODE 39 - ANSI MH10.8M-1983 - USD-3 - 3 of 9.
* C39+ : CODE 39 with checksum
* C39E : CODE 39 EXTENDED
* C39E+ : CODE 39 EXTENDED + CHECKSUM
* C93 : CODE 93 - USS-93
* S25 : Standard 2 of 5
* S25+ : Standard 2 of 5 + CHECKSUM
* I25 : Interleaved 2 of 5
* I25+ : Interleaved 2 of 5 + CHECKSUM
* C128 : CODE 128
* C128A : CODE 128 A
* C128B : CODE 128 B
* C128C : CODE 128 C
* EAN2 : 2-Digits UPC-Based Extention
* EAN5 : 5-Digits UPC-Based Extention
* EAN8 : EAN 8
* EAN13 : EAN 13
* UPCA : UPC-A
* UPCE : UPC-E
* MSI : MSI (Variation of Plessey code)
* MSI+ : MSI + CHECKSUM (modulo 11)
* POSTNET : POSTNET
* PLANET : PLANET
* RMS4CC : RMS4CC (Royal Mail 4-state Customer Code) - CBC (Customer Bar Code)
* KIX : KIX (Klant index - Customer index)
* IMB: Intelligent Mail Barcode - Onecode - USPS-B-3200
* CODABAR : CODABAR
* CODE11 : CODE 11
* PHARMA : PHARMACODE
* PHARMA2T : PHARMACODE TWO-TRACKS

2d barcodes
-----------
* DATAMATRIX : Datamatrix (ISO/IEC 16022)
* PDF417 : PDF417 (ISO/IEC 15438:2006) / a,e,t,s,f,o0,o1,o2,o3,o4,o5,o6
* QRCODE : QRcode Low error correction
* QRCODE, L : QRcode Low error correction
* QRCODE, M : QRcode Medium error correction
* QRCODE, Q : QRcode Better error correction
* QRCODE, H : QR-CODE Best error correction
* QR, RAW : raw mode - comma-separad list of array rows
* QR, RAW2 : raw mode - array rows are surrounded by square parenthesis.


Prerequisites
============

This version requires Symfony 2.1


Installation
============

  1 - Add the following lines in your composer.json:

  _for latest symfony 3.n kernel / framework environment:_
```
    "require": {
        "bitgrave/barcode-bundle": "^1.0"
    }
```
  _legacy symfony 2.1.n kernel? user version prior 1.0.0:
```
    "require": {
        "bitgrave/barcode-bundle": "0.0.9"
    }
```

  2 - Run the composer to download the bundle

```
    $ php composer.phar update bitgrave/barcode-bundle
```

  3 - Add this bundle to your application's kernel:

```
      // app/AppKernel.php
      public function registerBundles()
      {
          return array(
              // ...
              new BG\BarcodeBundle\BarcodeBundle(),
              // ...
          );
      }
```

Usage
=====

  1 - Add the following lines in your controller if you want to use both code types (1d/2d) :

```
use BG\BarcodeBundle\Util\Base1DBarcode as barCode;
use BG\BarcodeBundle\Util\Base2DBarcode as matrixCode;
```

  2 - set the cache path (for image based barcode rendering) call the image- or html renderer including your code and barcode type :

```
 $myBarcode = new barCode();
 $myBarcode->savePath = '/my/temp/media/path';
 $bcPathAbs = $myBarcode->getBarcodePNGPath('501234567890', 'EAN13', 1.75, 45);
 $bcHTMLRaw = $myBarcode->getBarcodeHTML('501234567890', 'EAN13', 1.75, 45);
```

 3 - fetch image by parse $bcPathAbs (absolute path to rendered barcode image) or using this a simple helper method

```
 /**
  * simple cache path returning method (sample cache path: "upload/barcode/cache" )
  *
  * @param bool $public
  *
  * @return string
  *
  */
 protected function getBarcodeCachePath($public = false)
 {

     return (!$public) ? $this->get('kernel')->getRootDir(). '/../web/upload/barcode/cache' : '/upload/barcode/cache';
 }
```

 4 - send public path to your symfony view and put result into your image src path or just render out the alternative table based barcode html structure

```
 $this->render('AcmeDemoBundle:Demo:barcode.html.twig', array(
     'barcodePathAndFile' => $this->getBarcodeCachePath($bcPathAbs),
     'barcodeHTML' => $bcHTMLRaw,
 ));

 // AcmeDemoBundle:Demo:barcode.html.twig
 // ...
 <!-- barcode as image -->
 <img src="{{ barcodePathAndFile }}" alt="barcode" title="my barcode image">
 // ...
 <!-- barcode as table -->
 {{ barcodeHTML|raw }}
 // ...

```


How To Contribute
=================

To contribute changes, fixes, additions/features please open a pull request with your new code.
please take not, that if you add new features or modify existing ones you have to doc this in
projects README file (also update projects CHANGELOG file!)

License
=======

See: resources/meta/LICENSE


Changelog.md
Changelog
=========

### 1.0.2 (23.06.2021)

#### Fixes
- unit tests
- compose json issue(s)
- core dependency updates


### 1.0.1 (24.01.2017)

#### Fixes
- unit tests
- minor doc issue
- phpunit definition file
- (re)set compatibility for php 5.5.n 


### 1.0.0 (24.01.2017)

#### Changes
- add support for latest symfony core


### 0.9.9 (31.08.2017)

#### Fixes
- merge pull request #19 from zittix/patch-1


### 0.9.8 (28.08.2017)

#### Fixes
- unit tests

#### Changes
- add/change copyright/licence model


Util/Base2DBarcode.php
<?php

//============================================================+
// File name   : 2dbarcodes.php
// Version     : 1.0.013
// Begin       : 2009-04-07
// Last Update : 2012-01-12
// Author      : Nicola Asuni - Tecnick.com LTD - Manor Coach House, Church Hill, Aldershot, Hants, GU12 4RQ, UK - www.tecnick.com - info@tecnick.com
// License     : GNU-LGPL v3 (http://www.gnu.org/copyleft/lesser.html)
// -------------------------------------------------------------------
// Copyright (C) 2009-2012  Nicola Asuni - Tecnick.com LTD
//
// This file is part of TCPDF software library.
//
// TCPDF is free software: you can redistribute it and/or modify it
// under the terms of the GNU Lesser General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// TCPDF is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
// See the GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with TCPDF.  If not, see <http://www.gnu.org/licenses/>.
//
// See LICENSE.TXT file for more information.
// -------------------------------------------------------------------
//
// Description : PHP class to creates array representations for
//               2D barcodes to be used with TCPDF.
//
//============================================================+

namespace BG\BarcodeBundle\Util;

use BG\BarcodeBundle\Util\Plugins\pdf417 as PDF417,
    BG\BarcodeBundle\Util\Plugins\datamatrix as Datamatrix,
    BG\BarcodeBundle\Util\Plugins\qrcode as QRcode;

/**
 * class Base2DBarcode 1.0.1, based on 2dbarcodes.php v1.0.013 (Nicola Asuni)
 *
 * @author Nicola Asuni
 * @author Dinesh Rabara <dinesh.rabara@gmail.com>
 * @author Patrick Paechnatz <patrick.paechnatz@gmail.com>
 */
class Base2DBarcode
{
    /**
     * Array representation of barcode.
     * @protected
     */
    protected $barcodeArray = [];

    /**
     * path to save png in getBarcodePNGPath
     * @var <type>
     */
    public $savePath;

    /**
     * Return an array representations of barcode.
     * @return array
     */
    public function getBarcodeArray()
    {
        return $this->barcodeArray;
    }

    /**
     * setup temporary path for barcode cache
     *
     * @param string $serverPath
     *
     * @throws \Exception
     */
    public function setTempPath($serverPath)
    {
        try {
            if (!file_exists($serverPath)) {
                mkdir($serverPath, 0770, true);
            }
        } catch (\Exception $e) {
            throw new \Exception("An error occurred while creating barcode cache directory at " . $serverPath);
        }
    }

    /**
     * get svg brcode (header stream)
     *
     * @param string $code
     * @param string $type
     * @param int    $w
     * @param int    $h
     * @param string $color
     */
    public function getBarcodeSVG($code, $type, $w=3, $h=3, $color='black')
    {
        //set barcode code and type
        $this->setBarcode($code, $type);
        // send headers
        $code = $this->getBarcodeSVGcode($w, $h, $color);
        header('Content-Type: application/svg+xml');
        header('Cache-Control: public, must-revalidate, max-age=0'); // HTTP/1.1
        header('Pragma: public');
        header('Expires: Sat, 12 Nov 1977 23:50:00 GMT'); // Date in the past
        header('Last-Modified: ' . gmdate('D, d M Y H:i:s') . ' GMT');
        header('Content-Disposition: inline; filename="' . md5($code) . '.svg";');
        //header('Content-Length: '.strlen($code));
        echo $code;
    }

    /**
     * Return a SVG string representation of barcode.
     *
     * @param string $code
     * @param string $type
     * @param int    $w
     * @param int    $h
     * @param string $color
     *
     * @return string
     */
    public function getBarcodeSVGcode($code, $type, $w=3, $h=3, $color='black')
    {
        //set barcode code and type
        $this->setBarcode($code, $type);
        // replace table for special characters
        $repstr = array("\0" => '', '&' => '&amp;', '<' => '&lt;', '>' => '&gt;');
        $svg = '<' . '?' . 'xml version="1.0" standalone="no"' . '?' . '>' . "\n";
        $svg .= '<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">' . "\n";
        $svg .= '<svg width="' . round(($this->barcodeArray['num_cols'] * $w), 3) . '" height="' . round(($this->barcodeArray['num_rows'] * $h), 3) . '" version="1.1" xmlns="http://www.w3.org/2000/svg">' . "\n";
        $svg .= "\t" . '<desc>' . strtr($this->barcodeArray['code'], $repstr) . '</desc>' . "\n";
        $svg .= "\t" . '<g id="elements" fill="' . $color . '" stroke="none">' . "\n";
        // print barcode elements
        $y = 0;
        // for each row
        for ($r = 0; $r < $this->barcodeArray['num_rows']; ++$r) {
            $x = 0;
            // for each column
            for ($c = 0; $c < $this->barcodeArray['num_cols']; ++$c) {
                if ($this->barcodeArray['bcode'][$r][$c] == 1) {
                    // draw a single barcode cell
                    $svg .= "\t\t" . '<rect x="' . $x . '" y="' . $y . '" width="' . $w . '" height="' . $h . '" />' . "\n";
                }
                $x += $w;
            }
            $y += $h;
        }
        $svg .= "\t" . '</g>' . "\n";
        $svg .= '</svg>' . "\n";

        return $svg;
    }

    /**
     * Return an HTML representation of barcode.
     *
     * @param string $code
     * @param string $type
     * @param int    $w
     * @param int    $h
     * @param string $color
     *
     * @return string
     */
    public function getBarcodeHTML($code, $type, $w=10, $h=10, $color='black')
    {
        //set barcode code and type
        $this->setBarcode($code, $type);
        $html = '<div style="font-size:0;position:relative;width:' . ($w * $this->barcodeArray['num_cols']) . 'px;height:' . ($h * $this->barcodeArray['num_rows']) . 'px;">' . "\n";
        // print barcode elements
        $y = 0;
        // for each row
        for ($r = 0; $r < $this->barcodeArray['num_rows']; ++$r) {
            $x = 0;
            // for each column
            for ($c = 0; $c < $this->barcodeArray['num_cols']; ++$c) {
                if ($this->barcodeArray['bcode'][$r][$c] == 1) {
                    // draw a single barcode cell
                    $html .= '<div style="background-color:' . $color . ';width:' . $w . 'px;height:' . $h . 'px;position:absolute;left:' . $x . 'px;top:' . $y . 'px;">&nbsp;</div>' . "\n";
                }
                $x += $w;
            }
            $y += $h;
        }
        $html .= '</div>' . "\n";

        return $html;
    }

    /**
     * Return an HTML representation of the barcode as a table, optimised as
     * much as possible for use in an email client (i.e. using legacy HTML
     * attributes and omitting any optional tags). Not XHTML compatible.
     * Only suitable for bi-color barcodes due to aggresive optimisation.
     *
     * @param string $code
     * @param string $type
     * @param int    $w
     * @param int    $h
     * @param string $color The CSS color used to represent 'black' bars/cells in the barcode
     * @param bool   $useColspan Use colspan to consolidate consecutive identical cells
     * @param bool   $emptyCells Non-empty cells are required for various versions of Outlook on Windows
     *
     * @return string
     */
    public function getBarcodeHTMLTable($code, $type, $w=10, $h=10, $color='#000', $useColspan=true, $emptyCells=false)
    {
        $this->setBarcode($code, $type);

        // reprocess the barcode data to allow the use of colspan when generating the HTML table
        $grid = [];
        for ($row = 0; $row < $this->barcodeArray['num_rows']; $row++) {
            $grid[$row] = [];

            $i = 0;
            for ($col = 0; $col < $this->barcodeArray['num_cols']; $col++) {
                if ($useColspan) {
                    if (isset($grid[$row][$i])) {
                        if ($this->barcodeArray['bcode'][$row][$col] == $this->barcodeArray['bcode'][$row][$col - 1]) {
                            $grid[$row][$i]['span']++;
                        } else {
                            $i++;
                        }
                    }
                } else {
                    $i = $col;
                }

                if (!isset($grid[$row][$i])) {
                    $grid[$row][$i] = [
                        'color' => $this->barcodeArray['bcode'][$row][$col] ? $color : '#fff',
                        'span' => 1,
                    ];
                }
            }
        }

        $width = $w * $this->barcodeArray['num_cols'];
        $html = '<table cellspacing=0 cellpadding=0 border=0';
        $html .= ' width='.$width;
        $html .= ' height='.($h * $this->barcodeArray['num_rows']);
        $html .= ' class=barcode';
        // font-size:1px is required for Outlook 2013, even if emptyCells is true
        $html .= ' style="font-size:1px;font-size:0;line-height:0;overflow:hidden;min-width:'.$width.'px;"';
        $html .= '>';

        if ($useColspan) {
            // a row of empty TDs mitigates the case where every cell in a column
            // is part of a colspan (otherwise the column collapses away)
            // COL elements would be ineffective, as Outlook ignores the width specified on them
            $html .= '<tr>';
            for ($i = 0; $i < $this->barcodeArray['num_cols']; $i++) {
                $html .= '<td width='.$w.' height=0>'.($emptyCells ? '' : '&#8203;');
            }
            $html .= '<tbody>';
        }

        foreach ($grid as $i => $row) {
            $html .= '<tr>';
            foreach ($row as $j => $col) {
                $html .= '<td';
                $html .= ($useColspan && $col['span'] > 1) ? ' colspan='.$col['span'] : '';
                // uses border-top rather than bgcolor so that the barcode prints when backgrounds are turned off
                $html .= ($col['color'] != '#fff' || ($j == 0 && count($row) == 1)) ?
                    ' style="border-top:'.$h.'px solid '.$col['color'].'"' : '';
                $html .= '>'.($emptyCells ? '' : '&#8203;'); // U+200B zwsp rather than nbsp to improve rendering
            }
        }
        $html .= '</table>';

        if ($useColspan && $emptyCells) {
            // colspan only saves bytes when spanning more than 3 empty white columns
            $html = str_replace('<td colspan=2>', '<td><td>', $html);
            $html = str_replace('<td colspan=3>', '<td><td><td>', $html);
        }

        return $html;
    }

    /**
     * Return a PNG image representation of barcode (requires GD or Imagick library).
     *
     * @param string $code
     * @param string $type
     * @param int    $w
     * @param int    $h
     * @param array  $color
     *
     * @return bool
     */
    public function getBarcodePNG($code, $type, $w=3, $h=3, $color=array(0, 0, 0))
    {
        //set barcode code and type
        $this->setBarcode($code, $type);
        $bar = null;

        // calculate image size
        $width = ($this->barcodeArray['num_cols'] * $w);
        $height = ($this->barcodeArray['num_rows'] * $h);
        if (function_exists('imagecreate')) {
            // GD library
            $imagick = false;
            $png = imagecreate($width, $height);
            $bgcol = imagecolorallocate($png, 255, 255, 255);
            imagecolortransparent($png, $bgcol);
            $fgcol = imagecolorallocate($png, $color[0], $color[1], $color[2]);
        } elseif (extension_loaded('imagick')) {
            $imagick = true;
            $fgcol = new \imagickpixel('rgb(' . $color[0] . ',' . $color[1] . ',' . $color[2] . ')');
            $png = new \Imagick();
            $png->newImage($width, $height, 'none', 'png');

        } else {

            return false;
        }
        // print barcode elements
        $y = 0;
        // for each row
        for ($r = 0; $r < $this->barcodeArray['num_rows']; ++$r) {
            $x = 0;
            // for each column
            for ($c = 0; $c < $this->barcodeArray['num_cols']; ++$c) {
                if ($this->barcodeArray['bcode'][$r][$c] == 1) {
                    // draw a single barcode cell
                    if ($imagick) {
                        $bar = new \imagickdraw();
                        $bar->setfillcolor($fgcol);
                        $bar->rectangle($x, $y, ($x + $w), ($y + $h));
                    } else {
                        imagefilledrectangle($png, $x, $y, ($x + $w), ($y + $h), $fgcol);
                    }
                }
                $x += $w;
            }
            $y += $h;
        }
        // send headers
        header('Content-Type: image/png');
        header('Cache-Control: public, must-revalidate, max-age=0'); // HTTP/1.1
        header('Pragma: public');
        header('Expires: Sat, 12 Nov 1977 23:50:00 GMT'); // Date in the past
        header('Last-Modified: ' . gmdate('D, d M Y H:i:s') . ' GMT');
        if ($imagick) {
            $png->drawimage($bar);
            echo $png;
        } else {
            imagepng($png);
            imagedestroy($png);
        }

        return true;
    }

    /**
     * return filename from give path
     *
     * @param string $path
     *
     * @todo: refactor this, move method in some kind of utility class
     *
     * @return mixed
     */
    public function getBarcodeFilenameFromGenPath($path)
    {
        $bcPathArr = explode('/', $path);

        return $bcPathArr[count($bcPathArr)-1];
    }

    /**
     * Return a .png file path which create in server
     *
     * @param string $code
     * @param string $type
     * @param int    $w
     * @param int    $h
     * @param array  $color
     *
     * @throws \Exception
     *
     * @return bool
     */
    public function getBarcodePNGPath($code, $type, $w=3, $h=3, $color=array(0, 0, 0),$filename = null)
    {
        if(is_null($filename)){
            $filename = $type.'_'.$code;
        }

        //set barcode code and type
        $this->setBarcode($code, $type);
        $bar = null;

        if (empty($this->barcodeArray) || (!$this->barcodeArray)) {
            throw new \Exception('It not possible to generate barcode of type: '.$type.' for number/code: '.$code.'! May be this is an invalid code pattern!');
        }

        // calculate image size
        $width = ($this->barcodeArray['num_cols'] * $w);
        $height = ($this->barcodeArray['num_rows'] * $h);
        if (function_exists('imagecreate')) {
            // GD library
            $imagick = false;
            $png = imagecreate($width, $height);
            $bgcol = imagecolorallocate($png, 255, 255, 255);
            imagecolortransparent($png, $bgcol);
            $fgcol = imagecolorallocate($png, $color[0], $color[1], $color[2]);
        } elseif (extension_loaded('imagick')) {
            $imagick = true;
            $fgcol = new \imagickpixel('rgb(' . $color[0] . ',' . $color[1] . ',' . $color[2] . ')');
            $png = new \Imagick();
            $png->newImage($width, $height, 'none', 'png');
        } else {

            return false;
        }
        // print barcode elements
        $y = 0;
        // for each row
        for ($r = 0; $r < $this->barcodeArray['num_rows']; ++$r) {
            $x = 0;
            // for each column
            for ($c = 0; $c < $this->barcodeArray['num_cols']; ++$c) {
                if ($this->barcodeArray['bcode'][$r][$c] == 1) {
                    // draw a single barcode cell
                    if ($imagick) {
                        $bar = new \imagickdraw();
                        $bar->setfillcolor($fgcol);
                        $bar->rectangle($x, $y, ($x + $w - 1), ($y + $h - 1));
                    } else {
                        imagefilledrectangle($png, $x, $y, ($x + $w - 1), ($y + $h - 1), $fgcol);
                    }
                }
                $x += $w;
            }
            $y += $h;
        }

        $this->setTempPath($this->savePath);
        $saveFile = $this->checkfile($this->savePath . $filename . '.png', true);

        if ($imagick) {
            $png->drawimage($bar);
        }

        if (imagepng($png, $saveFile)) {
            imagedestroy($png);

            return $saveFile;
        } else {
            imagedestroy($png);
            throw new \Exception('It not possible to write barcode cache file to path '.$this->savePath);
        }
    }

    /**
     * Set the barcode.
     *
     * @param string $code
     * @param string $type
     */
    public function setBarcode($code, $type)
    {
        $mode = explode(',', $type);
        $qrtype = strtoupper($mode[0]);
        switch ($qrtype) {
            case 'DATAMATRIX': // DATAMATRIX (ISO/IEC 16022)
                $qrcode = new Datamatrix($code);
                $this->barcodeArray = $qrcode->getBarcodeArray();
                $this->barcodeArray['code'] = $code;
                break;

            case 'PDF417': // PDF417 (ISO/IEC 15438:2006)
                if (!isset($mode[1]) || ($mode[1] === '')) {
                    $aspectratio = 2; // default aspect ratio (width / height)
                } else {
                    $aspectratio = floatval($mode[1]);
                }
                if (!isset($mode[2]) || ($mode[2] === '')) {
                    $ecl = -1; // default error correction level (auto)
                } else {
                    $ecl = intval($mode[2]);
                }
                // set macro block
                $macro = array();
                if (isset($mode[3]) && ($mode[3] !== '') && isset($mode[4]) && ($mode[4] !== '') && isset($mode[5]) && ($mode[5] !== '')) {
                    $macro['segment_total'] = intval($mode[3]);
                    $macro['segment_index'] = intval($mode[4]);
                    $macro['file_id'] = strtr($mode[5], "\xff", ',');
                    for ($i = 0; $i < 7; ++$i) {
                        $o = $i + 6;
                        if (isset($mode[$o]) && ($mode[$o] !== '')) {
                            // add option
                            $macro['option_' . $i] = strtr($mode[$o], "\xff", ',');
                        }
                    }
                }
                $qrcode = new PDF417($code, $ecl, $aspectratio, $macro);
                $this->barcodeArray = $qrcode->getBarcodeArray();
                $this->barcodeArray['code'] = $code;
                break;

            case 'QRCODE': // QR-CODE
                if (!isset($mode[1]) || (!in_array($mode[1], array('L', 'M', 'Q', 'H')))) {
                    $mode[1] = 'L'; // Ddefault: Low error correction
                }
                $qrcode = new QRcode($code, strtoupper($mode[1]));
                $this->barcodeArray = $qrcode->getBarcodeArray();
                $this->barcodeArray['code'] = $code;
                break;

            case 'RAW':
            case 'RAW2': // RAW MODE
                // remove spaces
                $code = preg_replace('/[\s]*/si', '', $code);
                if (strlen($code) < 3) {
                    break;
                }
                if ($qrtype == 'RAW') {
                    // comma-separated rows
                    $rows = explode(',', $code);
                } else {
                    // rows enclosed in square parentheses
                    $code = substr($code, 1, -1);
                    $rows = explode('][', $code);
                }
                $this->barcodeArray['num_rows'] = count($rows);
                $this->barcodeArray['num_cols'] = strlen($rows[0]);
                $this->barcodeArray['bcode'] = array();
                foreach ($rows as $r) {
                    $this->barcodeArray['bcode'][] = str_split($r, 1);
                }
                $this->barcodeArray['code'] = $code;
                break;

            default:
                $this->barcodeArray = false;
                break;
        }
    }

    /**
     * unlink old barcode image file or optional rand prefix file
     *
     * @param string $path
     * @param bool   $overwrite
     *
     * @return mixed
     */
    public function checkfile($path, $overwrite)
    {
        if (file_exists($path)) {

            if (!$overwrite) {
                $baseName = pathinfo($path, PATHINFO_BASENAME);

                return $this->checkfile(str_replace($baseName, rand(0, 9999) . $baseName, $path), $overwrite);

            } else {

                unlink($path);

            }
        }

        return $path;
    }
}
